README.md

```diff
@@ -1,21 +1,121 @@
-[![Dependencies](https://tidelift.com/badges/github/packagist/roave%2Fyou-are-using-it-wrong)](https://tidelift.com/subscription/pkg/packagist-roave%2Fyou-are-using-it-wrong?utm_source=packagist-ocramius%2Fpackage-versions&utm_medium=readme)
+
+This package enforces type checks during composer installation in downstream
+consumers of your package. This only applies to usages of classes, properties, methods and functions declared within packages that directly depend on *you-are-using-it-wrong*.  Issues that the static analyser finds that do not relate to these namespaces will not be reported.
+
+`roave/you-are-using-it-wrong` comes with a zero-configuration out-of-the-box
+setup.
+
+The usage of this plugin is highly endorsed for authors of new PHP libraries
+who appreciate the advantages of static types.
+
+This project is built with the hope that libraries with larger user-bases will
+raise awareness of type safety (or current lack thereof) in the PHP ecosystem.
+
+As annoying as it might sound, it is not uncommon for library maintainers to
+respond to support questions caused by lack of type checks in downstream
+projects. In addition to that, relying more on static types over runtime checks,
+it is possible to reduce code size and maintenance burden by strengthening the
+API boundaries of a library.
 
 ### Installation
 
-TBD
+This package is designed to be installed as a dependency of PHP **libraries**.
 
-### Use-cases
+In your library, add it to your
+`composer.json`:
 
-TBD
+```sh
+composer require roave/you-are-doing-it-wrong
+```
 
-## Professional Support
+No further changes are needed for this tool to start operating as per its
+design, if your declared types are already reflecting your library requirements.
+
+Please also note that this should **not** be used in `"require-dev"`, but
+specifically in `"require"` in order for the type checks to be applied to
+downstream consumers of your code.
+
+### Examples
+
+Assuming you have a library with following `composer.json`:
+
+```json
+{
+    "name": "my/awesome-library",
+    "type": "library",
+    "autoload": {
+        "psr-4": {
+            "My\\AwesomeLibrary\\": "src"
+        }
+    },
+    "require": {
+        "roave/you-are-doing-it-wrong": "@STABLE"
+    }
+}
+```
+
+Given following `src/MyHelloWorld.php`:
+
+```php
+<?php
 
+namespace My\AwesomeLibrary;
+
+final class MyHelloWorld
+{
+    public static function sayHello(string $name) : string
+    {
+        return 'Hello ' . $name . '!';
+    }
+}
+```
+
+Considering following downstream project:
+
+```json
+{
+    "name": "a/project",
+    "autoload": {
+        "psr-4": {
+            "The\\Project\\": "src"
+        }
+    }
+}
+```
+
+And following `src/MyExample.php`:
+
+```php
+<?php
+
+// notice the simple type error
+echo \My\AwesomeLibrary\MyHelloWorld::sayHello(123);
+```
+
+Then following command will fail with a type check error and description:
```

Comment:
This example bugs me to be honest. It says first to write a line of code with a library and then add it as a dependency. While this way of writing code is certainly possible, I don't think anyone writes code like this. I'd expect first to add in the dependency and then write the code using it.

With that in mind, feels like this plugin will trigger when adding whatever next dependency to my project.

Reply:
My guess on this one is that it is added to an existing library, and end-users get an error when upgrading said package.

Reply:
The example is simplistic, and designed to make the violation visible to the reader. The design of this package will make it act on any `composer install` or `composer update`, even for existing packages, or packages that are introduced via upgrades (indirectly).

Reply:
This section seems to spawn some doubts about this library. @BackEndTea is correct here in their assumptions: how can I change this section to be more obvious?

Reply:
Maybe just add something to that section to say that the check is also done on install and update. Hopefully most users will run composer commands much more often than they add new dependencies.

Reply:
I think any doubts would be cleared by having this example part of the documentation written something like this:
 - you (the reader) are the author of AwesomeLibrary, that uses strict types
 - you (the reader) add `you-are-using-it-wrong` as a dependency to AwesomeLibrary
 - Someone else decides to use your library, but they are using it wrong (wrong type passed to your amazing function)
 - On any/all composer calls this plugin will let them know that they are using your library wrong

Reply:
Added: 

 * runnable example
 * example output
 * clarified role (you = maintainer - not sure if sufficient)
 * used a clearer example with `array<string>`, which highlights the power of more generic types
 * clarified that any `composer install` and `composer update` will trigger this

Reply:
Much clearer, thanks! :+1: